lib/kennel/filter.rb

```diff
@@ -2,32 +2,45 @@
 
 module Kennel
   class Filter
+    attr_reader :project_filter, :tracking_id_filter
+
     def initialize
-      project_filter
-      tracking_id_filter
+      # build early so we fail fast on invalid user input
+      @tracking_id_filter = build_tracking_id_filter
+      @project_filter = build_project_filter
+    end
+
+    def filter_projects(projects)
+      filter_resources!(projects.dup, :kennel_id, project_filter, "projects", "PROJECT")
     end
 
-    def project_filter
-      projects = ENV["PROJECT"]&.split(",")&.sort&.uniq
-      tracking_projects = tracking_id_filter&.map { |id| id.split(":", 2).first }&.sort&.uniq
-      if projects && tracking_projects && projects != tracking_projects
+    def filter_parts(parts)
+      filter_resources!(parts.dup, :tracking_id, tracking_id_filter, "resources", "TRACKING_ID")
+    end
+
+    private
+
+    def build_project_filter
+      project_names = ENV["PROJECT"]&.split(",")&.sort&.uniq
+      tracking_project_names = tracking_id_filter&.map { |id| id.split(":", 2).first }&.sort&.uniq
+      if project_names && tracking_project_names && project_names != tracking_project_names
         raise "do not set PROJECT= when using TRACKING_ID="
       end
-      (projects || tracking_projects)
+      (project_names || tracking_project_names)
     end
 
-    def tracking_id_filter
+    def build_tracking_id_filter
       (tracking_id = ENV["TRACKING_ID"]) && tracking_id.split(",").sort.uniq
     end
 
-    def self.filter_resources!(resources, by, against, name, env)
-      return unless against
+    def filter_resources!(resources, by, against, name, env)
```

Comment:
To help readability. 

```suggestion
    def filter_resources!(resources, by, allow_list, name, env)
```